File: app/src/main/java/com/jmalinen/blackjack/engine/HandEvaluator.kt
package com.jmalinen.blackjack.engine

import com.jmalinen.blackjack.model.Card

object HandEvaluator {

    fun bestScore(cards: List<Card>): Int {
        if (cards.isEmpty()) return 0
        val aceCount = cards.count { it.rank.isAce }
        var sum = cards.filter { !it.rank.isAce }.sumOf { it.rank.baseValue }
        sum += aceCount // all aces as 1
        if (aceCount > 0 && sum + 10 <= 21) {
            sum += 10 // promote one ace to 11
        }
        return sum
    }

    fun isSoft(cards: List<Card>): Boolean {
        if (cards.none { it.rank.isAce }) return false
        val aceCount = cards.count { it.rank.isAce }
        val nonAceSum = cards.filter { !it.rank.isAce }.sumOf { it.rank.baseValue }
        return (nonAceSum + 11 + (aceCount - 1)) <= 21
    }
}

File: ios/Blackjack/Blackjack/Engine/HandEvaluator.swift
enum HandEvaluator {

    static func bestScore(_ cards: [Card]) -> Int {
        if cards.isEmpty { return 0 }
        let aceCount = cards.filter { $0.rank.isAce }.count
        var sum = cards.filter { !$0.rank.isAce }.reduce(0) { $0 + $1.rank.baseValue }
        sum += aceCount // all aces as 1
        if aceCount > 0 && sum + 10 <= 21 {
            sum += 10 // promote one ace to 11
        }
        return sum
    }

    static func isSoft(_ cards: [Card]) -> Bool {
        guard cards.contains(where: { $0.rank.isAce }) else { return false }
        let aceCount = cards.filter { $0.rank.isAce }.count
        let nonAceSum = cards.filter { !$0.rank.isAce }.reduce(0) { $0 + $1.rank.baseValue }
        return (nonAceSum + 11 + (aceCount - 1)) <= 21
    }
}
